### ISSUE TITLE
I hardly disagree on Question #305

### ISSUE THREAD

**Issue by iagocor** [knatten/cppquiz#240]
I kind of find it hard to disagree with this. But its marked as very easy so i am gonna go for it.
""""""
#include <stdio.h>
int main() {
    int i = 0;
    printf("%d%d", ++i, ++i);
    return 0;
}

Answer:  The program is guaranteed to output: 22 <- Wrong! its undefined (explicity, i think)

Explanation: §[intro.execution]¶18 When calling a function (whether or not the function is inline), every value computation and side effect associated with any argument expression, or with the postfix expression designating the called function, is sequenced before execution of every expression or statement in the body of the called function. 
"""""
It then proceeds to deep further into this and reproduce the steps of any path.
The problem is in the "reproduced path" after, in here particulary :
1. The leftmost[or rightmost] ++i is evaluated, which increments i to 1. The result of this expression is the updated i itself (not a copy).

-> printf in stdio takes C varargs, wich take by copy, thus the result of the evaluation is the rvalue "i" holds, 1, not a reference.

In godbolt i see that gcc gives the answer "22". This third option makes sense, because unsecuenced computations can be mixed, you can do both increments first and then both evaluations. 
However, this is changed in C++17(this, and that postfix are secuenced before any param, btw, kinda related) and now they must be secuenced, but its unspecified in what order. Yet, in gcc using --std==17 still prints 22, it appears to be wrong. Someone should call that out. Anyway, you can add another variable, a reference to i, to get a good result without changing the call.

Was this at some point specified behaviour? Am i missing something?

**Comment by knatten** [comment]
Thanks a lot for reporting!

I think it all comes down to poor wording in my explanation. This seems to be the core of your argument:

> -> printf in stdio takes C varargs, wich take by copy, thus the result of the evaluation is the rvalue "i" holds, 1, not a reference.

I agree with this. What I'm trying to say here is that the expression `++i` _itself_ refers to the object `i`. This is before we even get to passing that value to `printf`, I'm just talking about the result of that single expression. This is in contrast to `i++`, where the result is a prvalue, or a _copy_ of the original `i` as it was before it was incremented.

So the results of both those two `++i` expressions are `i` itself. See [expr.pre.incr](https://timsong-cpp.github.io/cppwp/n4861/expr.pre.incr):
> The result is the updated operand; it is an lvalue

After both of those side effects have completed, we now have two expressions which both _are_ the updated _i_. And now we can start thinking about calling `printf`, and by then it doesn't matter whether `printf` takes the arguments by value or reference.

Maybe I should change bullet point 3:

> `printf` is called with two arguments, both of which refer to `i`, which is now `2`.

The arguments _are_ both `i`, and using "refer" here might seem to indicate a reference. I was not pleased with that sentence when I wrote it, but I couldn't come up with something better. Would it help if I changed it to the following?

> `printf` is called with two arguments, both of which are `i`, which is now `2`.

Maybe I should also simply remove " (not a copy)", as this can be taken to mean that no copy is made for the function call, but what I actually mean is that the `++i` expression itself does not make a copy (like `i++` would have done). Or maybe change that bullet point to

> The leftmost ++i is evaluated, which increments i to 1. The result of this expression is the updated i itself (`++i` does not return a copy of `i` like `i++` does, instead it evaluates to `i` itself).

What do you think? Do you agree with my analysis?

**Comment by iagocor** [comment]
I understand your point, and the difference between pre and post increment is to take account, but i still think it doesn't matter here and that you are wrong in this: 
> After both of those side effects have completed, we now have two expressions which both are the updated i. And now we can start thinking about calling printf, 
TLDR: Not now! Anytime!
Computation of the operands [turning i into 1, putting it in place for the function to read] can be considered ""a side effect"", because it is also part of the evaluation of the lvalue expression "i", that in this case happens to be returned by the expression "++i".  I think they call it implicit lvalue to rvalue conversion.  The evaluation. along with side effects, is explicity mentioned in that quote from the standar, so... yeah. 
thus "and by then it doesn't matter whether printf takes the arguments by value or reference." is wrong. It does matter. If the function is passed a reference, then it sees the lvalue i and you are correct. If its passed a copy, it only see the rvalue of i but evaluated at any given time. Having only the preincrement return by reference doesn't work, if the "returned to" takes by value, Its like it returns by copy : reference + no reference = no reference
So i would change the example to using a function that takes i by reference. ¿? 

Consider this example: int i = 0; print ( i, ++i); . Its the same! And this is also unespecified! Or " i + (++i) ". Exactly the same! Because i is an lvalue, but it yet can give 0 (the result before the side effects of ++i) or i, after the side effects of i. 

printf (i++, i++) could legally devolve 2,2 prior to c++17 in the same way your example does. In fact, i think that the value of i could even be 1 at the end of the evaluation (0 + 1 computated twice), ¿?. Not after c++17 tho. 

And the example i said i added to godbolt 

https://godbolt.org/z/3M9sdMaYn

It shows even if you dont change the call to printf, just adding a declaration before the call works. Anyway, gcc is wrong, but besides that, acording to gcc its not garanteed to output 22 so it disagrees with you too. Gcc Its still wrong tho, 22 is ilegal in c++17 imo. But, anyway....

And i wouldn't change the "not a copy". Because thats the main point of the argument and its not wrong, but it shows the problem.

**Comment by knatten** [comment]
Oh, then I have misunderstood a key point here. I thought it would first evaluate all those expressions, then initialize the parameters from the result of each corresponding expression, then execute the function. But as you say, there are actually only two steps, not three. So the expression `++i` doesn't only compute increment `i`, it also initializes the parameter. It's obvious now that you point it out.

Clang actually prints `12` for the exact code in this question, as opposed to Gcc's non-conforming `22`: https://godbolt.org/z/aGrqdsW1r. Clang also warns about this by default, as does gcc with `-Wall`.

Another interesting thig: I modified your godbolt example a bit, to keep the _definition_ of `b`, but pass `i` for both arguments. And gcc still prints `21`! The mere existence of that reference happens to affect the output. Then if I delete that line, we're back to `22`. You can see all those three things here: https://godbolt.org/z/KzrsYEb8q.

I'll retract this question for now, and get back to improving it when I have more time. I'll leave the issue open in the mean time. I'll also email the author of the question and ask if they have any additional input.

**Comment by knatten** [comment]
I found the issue for this in GCC Bugzilla: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78734

I've written an updated explanation now, and would appreciate if you have time to review it. I also changed the question away from using `printf`, to avoid complicating the matters with a variadic function.

### New question:

```c++
#include <iostream>

void print(int x, int y)
{
    std::cout << x << y;
}

int main() {
    int i = 0;
    print(++i, ++i);
    return 0;
}
```

### New explanation:

The main question here is about the evaluation order of the two `++i` expressions, which initialize the two parameters `x` and `y` to `print`. `print` takes these by value, so a copy is made for each argument.

Two questions remain: In which order are the two `++i` expressions evaluated to initialize the parameters, and when do the increments of `i` take effect?

First of all, the evaluation order of the two `++i` expressions is unspecified. §[expr.call]¶5:
>  The initialization of a parameter, including every associated value computation and side effect, is indeterminately sequenced with respect to that of any other parameter.

What does "indeterminately sequenced" mean?

§[intro.execution]¶15:
> Evaluations A and B are indeterminately sequenced when either A is sequenced before B or B is sequenced before A, but it is unspecified which.

So from the two quotes above we see that they can happen in either order, and importantly that all side effects of the first evaluation (whichever that might be) are guaranteed to be completed before the second one gets evaluated.

Let's say the implementation decides to go left to right. Then the first parameter `x` is initialized by the leftmost expression `++i`, which increments `i` and results in the updated `i` which is now `1`. Next, the second parameter `y` is initialized by the rightmost expression `++i`, which increments `i` and results in the updated `i` which is now `2`. Then, `print` prints `12`.

Alternatively, the implementation could decide to go right to left, and `print` would print `21`.

So this is unspecified (but not undefined) behavior, and a conforming implementation can either print `12` or `21`, but not for instance `22`. Interestingly, `22` is what GCC prints at the time of writing, and there's an [open bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78734) for that.

**Comment by knatten** [comment]
Thank you so much for for reporting, explaining and reviewing! I've updated the question now and published it again: https://cppquiz.org/quiz/question/305
